apps/zipsync/src/zipSync.ts

```diff
@@ -0,0 +1,678 @@
+    );
+    markEnd('unpack.read.metadata');
+
+    terminal.writeLine(`Found ${entries.length} files in archive`);
+
+    for (const targetDirectory of targetDirectories) {
+      fs.mkdirSync(targetDirectory, { recursive: true });
+      terminal.writeDebugLine(`Ensured target directory: ${targetDirectory}`);
+    }
+
+    let extractedCount: number = 0;
+    let skippedCount: number = 0;
+    let deletedFilesCount: number = 0;
+    let deletedOtherCount: number = 0;
+    let deletedFoldersCount: number = 0;
+    let scanCount: number = 0;
+
+    const dirsToCleanup: string[] = [];
+
+    markStart('unpack.scan.existing');
+    const queue: IDirQueueItem[] = targetDirectories.map((dir) => ({
+      dir,
+      depth: 0,
+      node: zipTree.getNodeAtPrefix(path.relative(baseDir, dir))
+    }));
+
+    while (queue.length) {
+      const { dir: currentDir, depth, node } = queue.shift()!;
+      terminal.writeDebugLine(`Enumerating directory: ${currentDir}`);
+
+      const padding: string = depth === 0 ? '' : '-↳'.repeat(depth);
+
+      let items: fs.Dirent[];
+      try {
+        items = fs.readdirSync(currentDir, { withFileTypes: true });
+      } catch (e) {
+        terminal.writeWarningLine(`Failed to read directory: ${currentDir}`);
+        continue;
+      }
+
+      for (const item of items) {
+        scanCount++;
+        // check if exists in zipTree, if not delete
+        const relativePath: string = path
+          .relative(baseDir, path.join(currentDir, item.name))
+          .replace(/\\/g, '/');
+
+        const childNode: IReadonlyPathTrieNode<boolean> | undefined = node?.children?.get(item.name);
+
+        if (item.isFile()) {
+          terminal.writeVerboseLine(`${padding}${item.name}`);
+          if (!childNode?.value) {
+            terminal.writeDebugLine(`Deleting file: ${relativePath}`);
+            fs.unlinkSync(relativePath);
+            deletedFilesCount++;
+          }
+        } else if (item.isDirectory()) {
+          terminal.writeVerboseLine(`${padding}${item.name}/`);
+          queue.push({ dir: relativePath, depth: depth + 1, node: childNode });
+          if (!childNode || childNode.value) {
+            dirsToCleanup.push(relativePath);
+          }
+        } else {
+          terminal.writeVerboseLine(`${padding}${item.name} (not file or directory, deleting)`);
+          fs.unlinkSync(relativePath);
+          deletedOtherCount++;
+        }
+      }
+    }
+
+    for (const dir of dirsToCleanup) {
+      // Try to remove the directory. If it is not empty, this will throw and we can ignore the error.
+      try {
+        fs.rmdirSync(dir);
+        terminal.writeDebugLine(`Deleted empty directory: ${dir}`);
+        deletedFoldersCount++;
+      } catch (e) {
+        // Probably not empty
+        terminal.writeDebugLine(`Directory not empty, skipping: ${dir}`);
```

Comment:
This is possibly fatal. Also, log the underlying error. There are some we should check for.